backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java

```diff
@@ -0,0 +1,657 @@
+import ai.verta.modeldb.entities.versioning.BranchEntity;
+import ai.verta.modeldb.entities.versioning.CommitEntity;
+import ai.verta.modeldb.entities.versioning.RepositoryEntity;
+import ai.verta.modeldb.utils.ModelDBHibernateUtil;
+import ai.verta.uac.ModelResourceEnum;
+import ai.verta.uac.UserInfo;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.TimerTask;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+public class DeleteEntitiesCron extends TimerTask {
+  private static final Logger LOGGER = LogManager.getLogger(DeleteEntitiesCron.class);
+  private final AuthService authService;
+  private final RoleService roleService;
+  private static final String DATASET_GLOBAL_SHARING = "_DATASET_GLOBAL_SHARING";
+  private static final String REPOSITORY_GLOBAL_SHARING = "_REPO_GLOBAL_SHARING";
+
+  public DeleteEntitiesCron(AuthService authService, RoleService roleService) {
+    this.authService = authService;
+    this.roleService = roleService;
+  }
+
+  /** The action to be performed by this timer task. */
+  @Override
+  public void run() {
+    LOGGER.info("DeleteEntitiesCron wakeup");
+
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      // Update project timestamp
+      deleteProjects(session);
+
+      // Update experiment timestamp
+      deleteExperiments(session);
+
+      // Update experimentRun timestamp
+      deleteExperimentRuns(session);
+
+      // Update dataset timestamp
+      deleteDatasets(session);
+
+      // Update datasetVersion timestamp
+      deleteDatasetVersions(session);
+
+      // Update repository timestamp
+      deleteRepositories(session);
+    } catch (Exception ex) {
+      LOGGER.error("DeleteEntitiesCron Exception: ", ex);
+    }
+
+    LOGGER.info("DeleteEntitiesCron finish tasks and reschedule");
+  }
+
+  private void deleteProjects(Session session) {
+    LOGGER.debug("Project deleting");
+    String alias = "pr";
+    String deleteProjectsQueryString =
+        new StringBuilder("FROM ")
+            .append(ProjectEntity.class.getSimpleName())
+            .append(" ")
+            .append(alias)
+            .append(" WHERE ")
+            .append(alias)
+            .append(".")
+            .append(ModelDBConstants.DELETED)
+            .append(" = :deleted ")
+            .toString();
+
+    int lowerBound = 0;
+    final int pagesize = 500;
+    Query countQuery =
+        session.createQuery("SELECT COUNT(" + alias + ") " + deleteProjectsQueryString);
+    countQuery.setParameter("deleted", true);
+    Long count = (Long) countQuery.uniqueResult();
+    LOGGER.debug("Total projectEntities {} for deletion", count);
+
+    while (lowerBound < count) {
```

Comment:
This whole process seems pretty error prone. First you query the DB to get X project IDs, then you issue the deletion of each individually, then you issue update of experiment deletion with the X project IDs, then you delete the role bindings. And this sounds like it should be in a transaction if done in this way.

What about the following:
1. run request to get project IDs
2. delete role bindings. We do it at this step so that we can retry later if we get an error from UAC
3. run one request to update the experiments to set `deleted=1` based on the IDs we got before 
4. run one request to delete the projects based on the IDs we got before

In this order, no transactions are necessary since failure in any step before the last will leave the project with `deleted=1` there to be retried later.

The same logic is applied to the other ones below. The order of the steps remain the same. Always ask the question: if the system crashes at any particular step, can I restart safely? Right now, the answer is "no" unless we make a big transaction.

Side note: you don't need the loop with `lowerBound`. Just remove a configurable number of entries, just like we did on the update. Every time we'll get closer to the final state of having the things actually deleted, but we don't necessarily need to reach that in a single step, which might place a higher burden on the database.

Reply:
Change code as per the above suggestion